packages/grafana-data/src/transformations/transformers/seriesToRows.ts

```diff
@@ -70,13 +70,29 @@ export const seriesToRowsTransformer: DataTransformerInfo<SeriesToRowsTransforme
         for (let frameIndex = 0; frameIndex < data.length; frameIndex++) {
           const frame = data[frameIndex];
 
+          const firstNonTimeField = frame.fields[1];
+
+          // To support consistent naming for dataplane and prometheus
+          // we implement the following hierarchy:
+
+          // displayNameFromDS > frameDisplayName > fieldDisplayName
+
+          // This supports new naming and custom names (from the prom legend option)
+          // and supports the older pattern of getting the frame name.
+          // if neither of those return a name we use getFieldDisplayName which is good.
+          const displayNameFromDS = firstNonTimeField.config.displayNameFromDS;
+          const frameDisplayName = displayNameFromDS ?? getFrameDisplayName(frame);
+          const fieldDisplayName = getFieldDisplayName(firstNonTimeField, frame);
+
+          const displayName = displayNameFromDS ?? frameDisplayName ?? fieldDisplayName;
```

Comment:
when would frameDisplayName ever be null/undefined?

Reply:
If the data source is a backend data source and implements frames with NewFrame, name is required.

But, if there is a data source that has not been migrated to the backend and uses MutableDataFrame the name is optional.

(https://github.com/grafana/grafana/blob/0838c6168d85b65afd365c5bab8f3a13c24c32b9/packages/grafana-data/src/dataframe/MutableDataFrame.ts#L23-L27)

Reply:
The tsdb data sources that O11y metrics squad own, Prometheus, InfluxDB, Graphite and OpenTSDB all set the frame name.

Reply:
Oh, now I see, the fallback for getFieldDisplayName is here. This is already handled.

https://github.com/grafana/grafana/blob/a2f32a13bdf85e7b58bd1ad8685d06823f6547be/packages/grafana-data/src/field/fieldState.ts#L31-L37